src/lsm/table_memory.zig

```diff
@@ -163,37 +209,42 @@ pub fn TableMemoryType(comptime Table: type) type {
             assert(table.mutability == .mutable);
             assert(table.mutability.mutable.suffix_offset <= table.count());
 
-            table.sort_suffix_from_offset(table.mutability.mutable.suffix_offset);
+            table.mutable_sort_suffix_from_offset(table.mutability.mutable.suffix_offset);
 
             assert(table.mutability.mutable.suffix_offset == table.count());
         }
 
-        fn sort_suffix_from_offset(table: *TableMemory, offset: u32) void {
+        fn mutable_sort_suffix_from_offset(table: *TableMemory, offset: u32) void {
             assert(table.mutability == .mutable);
             assert(offset == table.mutability.mutable.suffix_offset or offset == 0);
             assert(offset <= table.count());
 
-            std.mem.sort(
-                Value,
-                table.values_used()[offset..],
-                {},
-                sort_values_by_key_in_ascending_order,
-            );
+            const target_count = sort_suffix_from_offset(table.values_used(), offset);
+            table.value_context.count = target_count;
+            table.mutability = .{ .mutable = .{ .suffix_offset = target_count } };
+        }
+
+        /// Returns the new length of `values`. (Values are deduplicated after sorting, so the
+        /// returned count may be less than or equal to the original `values.len`.)
+        fn sort_suffix_from_offset(values: []Value, offset: u32) u32 {
```

Comment:
👍 on passing values and offset directly, rather than `self`. I think that for hot data-plane methods that's the right choice, as it gives much more information to the optimizer about what can't change (and, also, just usually makes the code easier to follow, if a bit more verbose on the call site)